blocks/datasets/__init__.py

```diff
@@ -0,0 +1,233 @@
+            e.g. the number of examples to return, or the indices of a
+            particular minimbatch of examples.
+        sources : tuple of strings, optional
+            The data sources to return. By default all data sources are
+            returned.
+
+        .. todo::
+
+           A way for the dataset to communicate which kind of requests it
+           accepts, and a way to communicate what kind of request is being
+           sent when supporting multiple.
+
+        Returns
+        -------
+        tuple
+            A tuple of data.
+
+        """
+        raise NotImplementedError
+
+
+class DataStream(object):
+    """A stream of data.
+
+    A data stream implements the ``__iter__`` protocol, returning a new
+    iterator over the data each time it gets called (in the case of finite
+    data). A single pass over one of these iterators represents an *epoch*.
+
+    The data stream also opens a given dataset and maintains the state
+    (e.g. file handles) if necessary.
+
+    A data stream can provide an iterator over a dataset, but it can also
+    wrap another datastream e.g. to provide caching or some sort of
+    pre-processing.
+
+    Parameters
+    ----------
+    data : :class:`Dataset` or :class:`DataStream`
+        An object that implements the :meth:`open` and :meth:`get_data`
+        methods. Usually this is a :class:`Dataset` instance, but when
+        chaining iterators to e.g. perform caching this can also be a
+        :class:`DataStream` object.
+    iteration_scheme : :class:`IterationScheme`, optional
+        The iteration scheme to use when retrieving data. Note that not all
+        datasets support the same iteration schemes, some data sets require
+        one, and others don't support any.
+    sources : tuple of strings, optional
+        The sources of data to return. By default, all sources of the
+        dataset are requested.
+
+    """
+    def __init__(self, data, iteration_scheme=None, sources=None):
+        update_instance(self, locals())
+        self.data_state = None
+
+    def __iter__(self):
+        self.data_state = self.data.open(self.data_state)
+        # TODO Allow for multiple child datasets/data streams, so that they
+        # can be combined
+        if isinstance(self.data, DataStream):
+            self.child_iterator = iter(self.data)
+        return DataIterator(self, iter(self.iteration_scheme)
+                            if self.iteration_scheme else None, self.sources)
+
+    @property
+    def sources(self):
+        if getattr(self, '_sources', None) is None:
+            return self.data.sources
+
+    @sources.setter
+    def sources(self, sources):
+        self._sources = sources
+
+    def open(self, state=None):
+        return self.data.open(state)
+
+    def close(self, satte):
+        self.data.close(self.data_state)
+
+    def get_data(self, state=None, request=None, sources=None):
```

Comment:
Don't have the code in front of me but I believe it was to badge l make the
interface the same as Dataset, so that date streams can be agnostic as to
whether they are wrapping a data set or another data stream.
On Jan 9, 2015 9:34 AM, "Dmitry Bogdanov" <notifications@github.com> wrote:

> In blocks/datasets/__init__.py
> <https://github.com/bartvm/blocks/pull/91#discussion-diff-22705523>:
>
> > +    @property
> > +    def sources(self):
> > +        if getattr(self, '_sources', None) is None:
> > +            return self.data.sources
> > +
> > +    @sources.setter
> > +    def sources(self, sources):
> > +        self._sources = sources
> > +
> > +    def open(self, state=None):
> > +        return self.data.open(state)
> > +
> > +    def close(self, satte):
> > +        self.data.close(self.data_state)
> > +
> > +    def get_data(self, state=None, request=None, sources=None):
>
> Why do we need a state parameter if it is always self.data_state?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bartvm/blocks/pull/91/files#r22705523>.
>